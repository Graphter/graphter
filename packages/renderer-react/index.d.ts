import ListRenderer from './list-renderer';
export { ListRenderer };
