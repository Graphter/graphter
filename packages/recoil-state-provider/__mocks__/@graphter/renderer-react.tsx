import { NodeRendererRegistration } from "@graphter/core";
import React from "react";

export const nodeRendererStore = {
  get: jest.fn()
}

export const pathConfigStore = {
  get: jest.fn(),
  set: jest.fn()
}

export const NodeValidationProvider = jest.fn(({ children }: any) => <div>{children}</div>)
export const NodeDataProvider = jest.fn(({ children }: any) => <div>{children}</div>)