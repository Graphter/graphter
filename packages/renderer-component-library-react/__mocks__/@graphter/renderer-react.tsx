import React, { useState } from "react";

export const useArrayNodeData = jest.fn(() => {
  return {
    childIds: [],
    removeItem: jest.fn(),
    commitItem: jest.fn()
  }
})

export const useNodeData = jest.fn(() => {
  return [
    null,
    () => {}
  ]
})

export const useNodeValidation = jest.fn().mockResolvedValue([])

export const nodeRendererStore = {
  get: jest.fn((nodeType: string) => ({
    type: 'string',
    renderer: ({originalNodeData}: any) => {
      const [ value, setValue ] = useState(originalNodeData || '')
      return (
        <input value={value} onChange={e => setValue(e.target.value)} data-testid='string-renderer' />
    )
    }
  }))
}
export const createDefault = jest.fn()