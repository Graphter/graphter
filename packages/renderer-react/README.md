# Graphter React renderer

A React framework that, when combined with node renderers such as those provided in the [Component Lib](../renderer-component-library-react/README.md) package, allows humans to manage arbitrary data structures.