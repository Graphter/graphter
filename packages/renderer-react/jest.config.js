module.exports = {
  preset: 'ts-jest',
  testEnvironment: 'jsdom',
  moduleNameMapper: {
    "\\.(jpg|jpeg|png|gif|eot|otf|webp|svg|ttf|woff|woff2|mp4|webm|wav|mp3|m4a|aac|oga)$": "<rootDir>/__mocks__/fileMock.ts",
    "\\.(pcss|css|less)$": "identity-obj-proxy"
  },
  setupFilesAfterEnv: [ './setupTests.ts' ],
  automock: true,
  unmockedModulePathPatterns: [],
  snapshotSerializers: [
    'enzyme-to-json'
  ],
  testMatch: ['<rootDir>/**/*.test.ts*'],
  testPathIgnorePatterns: ['node_modules'],
};