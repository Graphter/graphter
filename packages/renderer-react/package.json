{
  "name": "malle-renderer-react",
  "version": "1.0.0",
  "description": "",
  "main": "dist/index.js",
  "module": "dist/index.es.js",
  "files": [
    "dist"
  ],
  "types": "dist/index.d.ts",
  "scripts": {
    "test": "cross-env CI=1 react-scripts test --env=jsdom",
    "test:watch": "react-scripts test --env=jsdom",
    "build": "rollup -c",
    "start": "rollup -c -w",
    "prepare": "npm run build"
  },
  "author": "",
  "license": "ISC",
  "peerDependencies": {
    "react": "^16.13.1",
    "react-dom": "^16.13.1"
  },
  "dependencies": {},
  "devDependencies": {
    "react": "^16.13.1",
    "react-dom": "^16.13.1",
    "@babel/core": "^7.9.0",
    "@babel/plugin-proposal-export-default-from": "^7.8.3",
    "@babel/preset-env": "^7.9.0",
    "@babel/preset-react": "^7.9.1",
    "@babel/preset-stage-0": "^7.8.3",
    "@svgr/rollup": "^5.3.0",
    "@types/react": "^16.9.25",
    "eslint": "^6.8.0",
    "react-scripts": "^3.4.1",
    "rollup": "^2.2.0",
    "rollup-plugin-babel": "^4.4.0",
    "rollup-plugin-commonjs": "^10.1.0",
    "rollup-plugin-node-resolve": "^5.2.0",
    "rollup-plugin-peer-deps-external": "^2.2.2",
    "rollup-plugin-postcss": "^2.5.0",
    "rollup-plugin-typescript2": "^0.26.0",
    "rollup-plugin-url": "^3.0.1",
    "typescript": "^3.8.3"
  }
}
