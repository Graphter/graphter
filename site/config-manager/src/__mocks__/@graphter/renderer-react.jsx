"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.pathConfigStore = exports.createDefault = exports.nodeRendererStore = exports.useNodeValidation = exports.useNodeData = exports.useArrayNodeData = void 0;
var react_1 = require("react");
exports.useArrayNodeData = jest.fn(function () {
    return {
        childIds: [],
        removeItem: jest.fn(),
        commitItem: jest.fn()
    };
});
exports.useNodeData = jest.fn(function () {
    return [
        null,
        function () { }
    ];
});
exports.useNodeValidation = jest.fn().mockResolvedValue([]);
exports.nodeRendererStore = {
    get: jest.fn(function (nodeType) { return ({
        type: 'string',
        renderer: function (_a) {
            var originalNodeData = _a.originalNodeData;
            var _b = react_1.useState(originalNodeData || ''), value = _b[0], setValue = _b[1];
            return (<input value={value} onChange={function (e) { return setValue(e.target.value); }} data-testid='string-renderer'/>);
        }
    }); })
};
exports.createDefault = jest.fn();
exports.pathConfigStore = {
    get: jest.fn(),
    set: jest.fn()
};
