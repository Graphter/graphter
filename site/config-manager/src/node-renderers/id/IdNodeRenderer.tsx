import React, { ComponentType, useEffect, useState } from "react";
import { NodeRendererProps } from "@graphter/core";
import { createDefault, useNodeData, useNodeValidation } from "@graphter/renderer-react";
import s from './IdNodeRenderer.module.css'
import { nanoid } from 'nanoid'
import { setupNodeRenderer } from "@graphter/renderer-react";

const filterRegExp = /[^a-z0-9-]/
const filterRegExpGlobal = /[^a-z0-9-]/g

const IdNodeRenderer: ComponentType<NodeRendererProps> = setupNodeRenderer((
  {
    config,
    originalNodeData,
    committed = true,
    path
  }: NodeRendererProps
) => {
  const isNew = typeof originalNodeData === 'undefined'
  if(isNew) originalNodeData = createDefault(config, '')
  const [ touched, setTouched ] = useState(false)
  const [ nodeData, setNodeData ] = useNodeData(path, config, originalNodeData, committed)
  const validationResults = useNodeValidation(path)
  const showFixButton = filterRegExp.test(nodeData)
  return (
    <>
      <input
        type='text'
        value={nodeData}
        data-nodetype='id'
        data-nodepath={path.join('/')}
        className={s.input}
        onChange={(e) => {
          if(!touched) setTouched(true)
          setNodeData && setNodeData(e.currentTarget.value);
        }} />
      {touched && (
        showFixButton ? (
          <button
            type='button'
            onClick={() => {
              setNodeData(nodeData.toLowerCase().replace(filterRegExpGlobal, '-'))
            }}
          >Fix</button>
        ) : (
          <span>Ok</span>
        )
      )}
      {touched &&
      validationResults.results.map((result, i) => (
        result.valid ? null : (
          <div className={s.error} key={i} data-testid='validation-error'>
            {result.errorMessage}
          </div>
        )
      ))}
    </>
  )
})

export default IdNodeRenderer