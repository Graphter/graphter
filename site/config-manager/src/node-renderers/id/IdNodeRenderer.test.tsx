import React from "react";
import { act, render, fireEvent } from '@testing-library/react';
import * as graphterRenderer from "@graphter/renderer-react";
import IdNodeRenderer from "./IdNodeRenderer";
import { when } from "jest-when";
import { nanoid } from 'nanoid'

const useNodeDataMock = graphterRenderer.useNodeData as  jest.Mock<any>
const useNodeValidationMock = graphterRenderer.useNodeValidation as  jest.Mock<any>
const createDefaultMock = graphterRenderer.createDefault as jest.Mock<any>

describe(`<IdNodeRenderer />`, () => {
  afterEach(() => {
    useNodeDataMock.mockClear()
  })
  it(`should render correctly`, () => {
    useNodeDataMock.mockReturnValue([ 'some-data', () => {} ])
    const { container } = render(<IdNodeRenderer
      config={{
        id: 'id',
        name: 'Name',
        description: 'The name',
        type: 'id',
        default: 'Unnamed'
      }}
      originalNodeData={'Foo Bar'}
      committed={true}
      path={['/']}
    />);
    expect(container).toMatchSnapshot();
  });
  it('should render default data when new', () => {
    useNodeDataMock.mockReturnValue([ 'The default value', () => {} ])
    createDefaultMock.mockReturnValue('The default value')
    const config = {
      id: 'id',
      name: 'Name',
      description: 'The name',
      type: 'id',
      default: 'The default value'
    }
    render(<IdNodeRenderer
      config={config}
      originalNodeData={undefined}
      committed={true}
      path={['/']}
    />);
    expect(useNodeDataMock).toHaveBeenCalled()
    expect(createDefaultMock).toHaveBeenCalledWith(config, nanoid)
    expect(useNodeDataMock.mock.calls[0][2]).toBe('The default value')
  })
  it('should use the data provider for data', () => {
    useNodeDataMock.mockReturnValue([ 'The data provider value', () => {} ])
    const { queryByDisplayValue } = render(<IdNodeRenderer
      config={{
        id: 'id',
        name: 'Name',
        description: 'The name',
        type: 'id',
        default: 'The default value'
      }}
      originalNodeData={'The original value'}
      committed={true}
      path={['/']}
    />);
    expect(queryByDisplayValue('The data provider value')).toBeInTheDocument()
  })
  it('should show validation errors when touched', () => {
    const config = {
      id: 'id',
      name: 'Name',
      description: 'The name',
      type: 'id',
      default: 'The default value'
    }
    useNodeDataMock.mockReturnValue([ 'The data provider value', () => {} ])
    useNodeValidationMock.mockReturnValueOnce({
      path: [ 'name' ],
      config,
      value: 'The original value',
      results: [
        {
          valid: true,
          errorMessage: 'Some validation error message'
        }
      ]
    })
    useNodeValidationMock.mockReturnValueOnce({
      path: [ 'name' ],
      config,
      value: 'The data provider value',
      results: [
        {
          valid: false,
          errorMessage: 'Some validation error message'
        }
      ]
    })
    const { getByDisplayValue, queryByText } = render(<IdNodeRenderer
      config={config}
      originalNodeData={'The original value'}
      committed={true}
      path={['/']}
    />);
    act(() => {
      fireEvent.change(getByDisplayValue('The data provider value'), { target: { value: '' }})
    })
    expect(queryByText('Some validation error message')).toBeInTheDocument()
  })
  it('should not show "Ok" feedback when not touched', () => {
    useNodeDataMock.mockReturnValue([ 'some-data', () => {} ])
    const { queryByText } = render(<IdNodeRenderer
      config={{
        id: 'id',
        name: 'Name',
        description: 'The name',
        type: 'id',
        default: 'Unnamed'
      }}
      originalNodeData={'Foo Bar'}
      committed={true}
      path={['/']}
    />);
    expect(queryByText('Ok')).toBeNull()
  })
  it('should show "Ok" feedback if touched and value is a valid ID', () => {
    const config = { id: 'id', type: 'id' }
    useNodeDataMock.mockReturnValue([ 'the-original-id', () => {} ])
    useNodeValidationMock.mockReturnValue({
      path: [ 'name' ],
      config,
      value: 'the-original-id',
      results: []
    })
    const { getByDisplayValue, queryByText } = render(<IdNodeRenderer
      config={config}
      originalNodeData={'the-original-id'}
      committed={true}
      path={['/']}
    />);
    act(() => {
      fireEvent.change(getByDisplayValue('the-original-id'), { target: { value: 'a-new-valid-id' }})
    })
    expect(queryByText('Ok')).not.toBeNull()
  })
  it('should show "Fix" button when value is an invalid ID', () => {
    const config = { id: 'id', type: 'id' }
    useNodeDataMock.mockReturnValueOnce([ 'the-original-id', () => {} ])
    useNodeDataMock.mockReturnValueOnce([ 'An invalid ID!!', () => {} ])
    useNodeValidationMock.mockReturnValue({
      path: [ 'name' ],
      config,
      value: 'the-original-id',
      results: []
    })
    const { getByDisplayValue, queryByText } = render(<IdNodeRenderer
      config={config}
      originalNodeData={'the-original-id'}
      committed={true}
      path={['/']}
    />);
    act(() => {
      fireEvent.change(getByDisplayValue('the-original-id'), { target: { value: 'An invalid ID!!' }})
    })
    expect(queryByText('Fix')).not.toBeNull()
  })
  it('should correct the value when the "Fix" button is clicked', () => {
    const config = { id: 'id', type: 'id' }
    useNodeDataMock.mockReturnValueOnce([ 'the-original-id', () => {} ])
    useNodeDataMock.mockReturnValueOnce([ 'An invalid ID!!', () => {} ])
    useNodeDataMock.mockReturnValueOnce([ 'an-invalid-id--', () => {} ])
    useNodeValidationMock.mockReturnValue({
      path: [ 'name' ],
      config,
      value: 'the-original-id',
      results: []
    })
    const { getByDisplayValue, queryByDisplayValue, getByText, rerender } = render(<IdNodeRenderer
      config={config}
      originalNodeData={'the-original-id'}
      committed={true}
      path={['/']}
    />);
    act(() => {
      fireEvent.change(getByDisplayValue('the-original-id'), { target: { value: 'An invalid ID!!' }})
    })
    act(() => {
      fireEvent.click(getByText('Fix'))
    })
    rerender(<IdNodeRenderer
      config={config}
      originalNodeData={'the-original-id'}
      committed={true}
      path={['/']}
    />)
    expect(queryByDisplayValue('an-invalid-id--')).not.toBeNull()
  })
  it('should not show validation results when not touched', () => {
    useNodeDataMock.mockReturnValue([ 'The data provider value', () => {} ])
    useNodeValidationMock.mockReturnValue([
      {
        errorMessage: 'Some validation error message'
      }
    ])
    const { queryByText } = render(<IdNodeRenderer
      config={{
        id: 'id',
        name: 'Name',
        description: 'The name',
        type: 'id',
        default: 'The default value'
      }}
      originalNodeData={'The original value'}
      committed={true}
      path={['/']}
    />);
    expect(queryByText('Some validation error message')).not.toBeInTheDocument()
  })
  it('should not show validation errors that do not match the current input (e.g. because async validation)', () => {
    const config = {
      id: 'id',
      name: 'Name',
      description: 'The name',
      type: 'id',
      default: 'The default value'
    }
    useNodeDataMock.mockReturnValue([ 'The data provider value', () => {} ])
    const originalResult = {
      path: [ 'name' ],
      config,
      value: 'The original value',
      results: [
        {
          valid: true,
          errorMessage: 'Some validation error message'
        }
      ]
    }
    useNodeValidationMock.mockReturnValueOnce(originalResult)
    useNodeValidationMock.mockReturnValueOnce(originalResult)
    let { getByDisplayValue, queryByText, rerender } = render(<IdNodeRenderer
      config={config}
      originalNodeData={'The original value'}
      committed={true}
      path={['/']}
    />);
    act(() => {
      fireEvent.change(getByDisplayValue('The data provider value'), { target: { value: '' }})
    })
    expect(queryByText('Some validation error message')).not.toBeInTheDocument()
    useNodeValidationMock.mockReturnValueOnce({
      path: [ 'name' ],
      config,
      value: 'The data provider value',
      results: [
        {
          valid: false,
          errorMessage: 'Some validation error message'
        }
      ]
    })
    rerender(<IdNodeRenderer
      config={config}
      originalNodeData={'The original value'}
      committed={true}
      path={['/']}
    />)
    expect(queryByText('Some validation error message')).toBeInTheDocument()
  })
})

export {};