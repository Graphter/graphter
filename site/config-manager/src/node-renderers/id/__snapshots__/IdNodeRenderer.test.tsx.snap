// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IdNodeRenderer /> should render correctly 1`] = `
<div>
  <input
    class="input"
    data-nodepath="/"
    data-nodetype="id"
    type="text"
    value="some-data"
  />
</div>
`;
